text/3698-declarative-derive-macros.md

```diff
@@ -0,0 +1,144 @@
+- Feature Name: `declarative_derive_macros`
+- Start Date: 2024-09-20
+- RFC PR: [rust-lang/rfcs#3698](https://github.com/rust-lang/rfcs/pull/3698)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support implementing `derive(Trait)` via a `macro_rules!` macro.
+
+# Motivation
+[motivation]: #motivation
+
+Many crates support deriving their traits with `derive(Trait)`. Today, this
+requires defining proc macros, in a separate crate, typically with several
+additional dependencies adding substantial compilation time, and typically
+guarded by a feature that users need to remember to enable.
+
+However, many common cases of derives don't require any more power than an
+ordinary `macro_rules!` macro. Supporting these common cases would allow many
+crates to avoid defining proc macros, reduce dependencies and compilation time,
+and provide these macros unconditionally without requiring the user to enable a
+feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can define a macro to implement `derive(MyTrait)` by defining a
+`macro_rules!` macro with the `#[macro_derive]` attribute. Such a macro can
+create new items based on a struct, enum, or union. Note that the macro can
+only append new items; it cannot modify the item it was applied to.
+
+For example:
+
+```rust
+trait Answer { fn answer(&self) -> u32; }
+
+#[macro_derive]
+macro_rules! Answer {
+    // Simplified for this example
+    (struct $n:ident $_:tt) => {
```

Comment:
To be fair, it's very hard to parse all the variety of the Rust struct syntax using the macro matchers syntax, which makes this applicable in very-very simple cases. However, I **do** believe this feature will be useful, albeit in an other non-obvious way.

The fact that this macro has access to `$crate` basically solves the problem of discovering "home `crate`" for the regular proc macros. I myself already see using this feature to write a macro_rules-based derive macro wrapper for `derive(Builder)` like this:

```rust
#[macro_derive]
macro_rules! Builder {
    ($item:item) => {
        #[derive($crate::__BuilderImpl)]
        #[builder(crate = $crate)] // The proc-macro-based  `__BuilderImpl` derive macro now knows its home! 🎉 
    }
}
```
Where the `__BuilderImpl` is an internal proc-macro-based derive macro. In fact, this technique can also be applied to `serde`. `serde` can deprecate its [`crate` attribute](https://serde.rs/container-attrs.html#crate) entirely to remove that workaround.

I think such a use case deserves a couple of words in the RFC, WDYT?

Unfortunately, I don't see how this can be used to solve such a problem for proc-macro attributes 😿